### ISSUE TITLE
CA2249 encourages folks to ifdef code for "readability"

### ISSUE THREAD

**Issue by ericstj** [dotnet/roslyn-analyzers#4998]
### Analyzer

**Diagnostic ID**: [ca2249](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2249): `Consider using String.Contains instead of String.IndexOf`

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.100](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

The analyzer doesn't consider cross-compilation so it tells you to do something to improve readability and the only change that can be made is less readable (and less performant).

### Steps To Reproduce

Do the following in a project with 
```xml
<TargetFrameworks>net48;net50</TargetFrameworks>
```

```csharp
string toParse = "5";
if (toParse.IndexOf('.') == -1)
{
    toParse += ".0";
}
```

Observe the error:
```
Error	CA2249	Use 'string.Contains' instead of 'string.IndexOf' to improve readability
```

### Expected behavior

No warning since the recommendation can't be made consistently for all cross-targeting configurations.

### Actual behavior

A warning where the only way to remove it is to use a less performant string search, or to ifdef.

```csharp
#if NETCOREAPP
            if (!toParse.Contains('.'))
#else
            if (toParse.IndexOf('.') == -1)
#endif
```

**Comment by ericstj** [comment]
cc @pgovind

**Comment by mavasani** [comment]
Also tagging @sharwell and @jaredpar

This is not an issue specific to this analyzer, but is the inherent limitation of all analyzers and the compilation model of their execution. Analyzers only run in context of a single compilation at a time, and given each compilation for a project TFM can have different available APIs or different conditionally compiled code, they can always produce diagnostics that are applicable to a subset of TFMs. Unless there is some effort at MSBuild level to unify/filter out analyzer diagnostics that are not reported on all TFMs targeted by a project, there is nothing that can be done at the analyzer or compiler level here.

**Comment by sharwell** [comment]
The recommendation for a user in this situation would be either disabling this analyzer, or (if possible) providing a `string`-specific extension method that gets used for the net48. builds.
